themes/default/content/blog/five-years-of-infrastructure-as-code-part-one/index.md

```diff
@@ -0,0 +1,208 @@
+
+When we sat down to envision Pulumi, we had just been immersed in developer platforms for decades. We honestly had no idea we’d start with infrastructure as code. We tinkered with a new programming language, prototyped some higher level containers- and serverless-specific tools, frameworks, and programming models, … And yet, all roads kept coming back to infrastructure as code. It wasn’t that infrastructure as code was “bad” – indeed, it was essential – it was simply that the past attempts missed the mark. The “code” that these tools touted didn’t feel much like code to us at all in the usual sense. As a result, it didn’t have universal appeal, and most developers didn’t enjoy using these tools. In fact, it seemed the industry had taken a step back compared to the earliest infrastructure as code tools like Chef and Puppet that embraced languages. We realized that we could bring great programmability to all things modern cloud, opening it up to a universal audience.
+
+Why did all roads lead back to infrastructure as code? Infrastructure as code offers a fundamental building block that gives us a highly programmable and composable surface area over the entirety of the modern cloud’s capabilities. Being able to abstract and encapsulate complexity, and build bigger things out of smaller things, is the key not just to productivity, but also many decades of prior success and progress in computer science. Infrastructure as code also introduces a necessary concept of cloud resource lifetime and management that we take for granted in other areas of programming, for instance, with in-memory or operating system resources. Our “ah-hah” moment was to take everything we knew and loved about general-purpose languages, and marry that with infrastructure as code. In short, we set out to stand on the shoulders of giants and re-imagine infrastructure as code from start to finish.
+
+Let’s take a more in-depth look at that journey and how we ended up with Pulumi.
+
+## The Cloud as the New Operating System
+
+Most “code” we’ve written historically runs on a single computer. That single computer has resources like CPU, memory, disks, and network cards that the operating system manages. The lifetime of that hardware is relatively straightforward since programs have a finite lifetime. There are well defined process models where a program begins and ends running. Most resource usage is ephemeral and lasts no longer than a program’s lifetime, with the obvious exception of persistent state such as what gets written to the filesystem.
```

Comment:
```suggestion
Most “code” we’ve written historically runs on a single computer. That single computer has resources like CPU, memory, disks, and network cards that the operating system manages. The lifetime of those hardware resources is relatively straightforward since programs have a finite lifetime. There are well defined process models where a program begins and ends running. Most resource usage is ephemeral and lasts no longer than a program’s lifetime, with the obvious exception of persistent state such as what gets written to the filesystem.
```